devolutions-crypto-wayk/src/bastion/fle.rs

```diff
@@ -0,0 +1,203 @@
+    let (data, rest) = rest.split_at_mut(rest.len() - TAG_SIZE);
+    data[..content.len()].copy_from_slice(&content);
+    Pkcs7::pad(data, content.len(), block_size).map_err(|_| Error::Padding)?;
+
+    // in-place encryption
+    let key = Key::from_slice(key);
+    let aead = XChaCha20Poly1305::new(key);
+    let tag = aead
+        .encrypt_in_place_detached(&nonce, aad, data)
+        .map_err(|_| Error::XChaCha20)?;
+
+    // finally, write tag
+    rest.copy_from_slice(&tag);
+
+    // our final payload (ie: |~Safe~|k1AjQZYciYvlPmJMnHjbb…)
+    let mut final_payload = String::from(SAFE_PAYLOAD_PREFIX);
+    base64::encode_config_buf(&buffer, base64::STANDARD, &mut final_payload);
+
+    Ok(final_payload)
+}
+
+/// Decrypts field-level content.
+///
+/// - Input is the encrypted key encoded in base64.
+/// - Master Key ID is used for early key mismatch detection.
+/// - Output is decrypted field content.
+///
+/// # Interals
+///
+/// - Decryption is done using XChaCha2020-Poly1035.
+/// - Data layout is assumed to be the same as described in `encrypt_field_with_block_size`.
+pub fn decrypt_field(
+    master_key_id: Uuid,
+    key: &[u8],
+    encrypted_content: &str,
+) -> Result<Vec<u8>, Error> {
+    use chacha20poly1305::aead::Tag;
+
+    const MIN_SIZE: usize = AAD_SIZE + IV_SIZE + 1 + TAG_SIZE;
+
+    let mut buffer = base64::decode_config(
+        &encrypted_content[SAFE_PAYLOAD_PREFIX.len()..],
+        base64::STANDARD,
+    )?;
+
+    if buffer.len() < MIN_SIZE {
+        return Err(Error::InvalidSize { got: buffer.len() });
+    }
+
+    let (aad, rest) = buffer.as_mut_slice().split_at_mut(AAD_SIZE);
+    {
+        // associated data sanity checks
+
+        let (aad_magic, aad_rest) = aad.split_at(AAD_MAGIC_NUMBER_SIZE);
+        if aad_magic != AAD_MAGIC_NUMBER {
+            return Err(Error::InvalidMagicNumber);
+        }
+
+        let (_aad_reserved, uuid) = aad_rest.split_at(AAD_RESERVED_SIZE);
+        let found_master_key_id = Uuid::from_bytes([
+            uuid[0], uuid[1], uuid[2], uuid[3], uuid[4], uuid[5], uuid[6], uuid[7], uuid[8],
+            uuid[9], uuid[10], uuid[11], uuid[12], uuid[13], uuid[14], uuid[15],
+        ]);
+        if master_key_id != found_master_key_id {
+            return Err(Error::MasterKeyIdMismatch {
+                expected: master_key_id,
+                found: found_master_key_id,
+            });
+        }
+    }
+
+    let (nonce, rest) = rest.split_at_mut(IV_SIZE);
+    let nonce = XNonce::from_slice(nonce);
+
+    let (data, tag) = rest.split_at_mut(rest.len() - TAG_SIZE);
+
+    let tag = Tag::<<XChaCha20Poly1305 as AeadInPlace>::TagSize>::from_slice(tag);
+
+    let key = Key::from_slice(key);
+    let aead = XChaCha20Poly1305::new(key);
+    aead.decrypt_in_place_detached(nonce, aad, data, tag)
+        .map_err(|_| Error::XChaCha20)?;
+
+    // unpad data
+    let decoded_payload = Pkcs7::unpad(data).map_err(|_| Error::Padding)?;
+
+    Ok(decoded_payload.to_vec())
```

Comment:
We might want to zeroize data and decoded_payload before returning (and after it has been converted to a Vec)

Reply:
`decoded_payload` is a view on `data` so I'll only zeroize `data` here